lxd/db/cluster/identities.go

```diff
@@ -0,0 +1,214 @@
+//
+//go:generate mapper stmt -e identity objects table=identities
+//go:generate mapper stmt -e identity objects-by-ID table=identities
+//go:generate mapper stmt -e identity objects-by-AuthMethod table=identities
+//go:generate mapper stmt -e identity objects-by-AuthMethod-and-Type table=identities
+//go:generate mapper stmt -e identity objects-by-AuthMethod-and-Identifier table=identities
+//go:generate mapper stmt -e identity objects-by-AuthMethod-and-Name table=identities
+//go:generate mapper stmt -e identity objects-by-Type table=identities
+//go:generate mapper stmt -e identity id table=identities
+//go:generate mapper stmt -e identity create struct=Identity table=identities
+//go:generate mapper stmt -e identity delete-by-AuthMethod-and-Identifier table=identities
+//go:generate mapper stmt -e identity delete-by-Name-and-Type table=identities
+//go:generate mapper stmt -e identity update struct=Identity table=identities
+//
+//go:generate mapper method -i -e identity GetMany
+//go:generate mapper method -i -e identity GetOne
+//go:generate mapper method -i -e identity ID struct=Identity
+//go:generate mapper method -i -e identity Exists struct=Identity
+//go:generate mapper method -i -e identity Create struct=Identity
+//go:generate mapper method -i -e identity DeleteOne-by-AuthMethod-and-Identifier
+//go:generate mapper method -i -e identity DeleteMany-by-Name-and-Type
+//go:generate mapper method -i -e identity Update struct=Identity
+
+// AuthMethod is a database representation of an authentication method.
+type AuthMethod string
+
+const (
+	authMethodTLS  int64 = 1
+	authMethodOIDC int64 = 2
+)
+
+// Scan implements sql.Scanner for AuthMethod. This converts the integer value back into the correct API constant or
+// returns an error.
+func (a *AuthMethod) Scan(value any) error {
+	if value == nil {
+		return fmt.Errorf("Authentication method cannot be null")
+	}
+
+	intValue, err := driver.Int32.ConvertValue(value)
+	if err != nil {
+		return fmt.Errorf("Invalid authentication method type: %w", err)
+	}
+
+	authMethodInt, ok := intValue.(int64)
+	if !ok {
+		return fmt.Errorf("Authentication method should be an integer, got \"%v\" (%T)", intValue, intValue)
+	}
+
+	switch authMethodInt {
+	case authMethodTLS:
+		*a = api.AuthenticationMethodTLS
+	case authMethodOIDC:
+		*a = api.AuthenticationMethodOIDC
+	default:
+		return errors.New("Unknown authentication method")
+	}
+
+	return nil
+}
+
+// Value implements driver.Valuer for AuthMethod. This converts the API constant into an integer or throws an error.
+func (a AuthMethod) Value() (driver.Value, error) {
+	switch a {
+	case api.AuthenticationMethodTLS:
+		return authMethodTLS, nil
+	case api.AuthenticationMethodOIDC:
+		return authMethodOIDC, nil
+	}
+
+	return nil, fmt.Errorf("Invalid authentication method %q", a)
+}
+
+// IdentityType indicates the type of the identity.
+type IdentityType string
+
+const (
+	identityTypeRestrictedClientCertificate   int64 = 1
```

Comment:
For DB constants I tend to prefer them being explicitly defined so that ordering doesnt change the value.

Reply:
While it's less efficient, I often prefer strings because they're easier to log and debug, and you avoid the eventual need to write a `String()` function.

Reply:
in this case we are storing these into the db so they should be ints, and we will likely need a conversion function as you say. If it was just for the API string constants would be fine.

Reply:
A bit of confusion here so I will add more comments. The gist is that the `Scan` and `Value` functions defined on `IdentityType` and `AuthMethod` are performing the conversion from the string constants defined in `shared/api/auth.go` to the integers defined here and vice versa (that's why they're `int64`, to be compatible with the sqlite driver). This means that we can use the string constants (almost) everywhere in the codebase and even pass them into the generated methods directly (as they can be casted). If there is any conversion to do, it will just be a cast, e.g. `string(identity.AuthMethod)` or `cluster.AuthMethod(api.AuthenticationMethodTLS)`. See examples in `lxd/db/cluster/certificates.go`.

For me this is cleaner but perhaps it's preference.

Reply:
I have added comments to explain my reasoning for defining the `AuthMethod` and `IdentityType` types in this way. IMO this is a cleaner approach than doing conversions manually. Though I understand it's a little more obfuscated.